File: tutorials/a3-bayesian-inference/cxx/run-cmaes.cpp
// In this example, we demonstrate how Korali find the variable values
// that maximize the posterior in a bayesian problem where the likelihood
// is calculated by providing reference data points and their objective values.

#include "korali.hpp"
#include "model/model.h"

int main(int argc, char* argv[])
{
 auto k = Korali::Engine();

 // Selecting problem and likelihood model
 k["Problem"]["Type"] = "Bayesian Inference";
 k["Problem"]["Likelihood"]["Model"] = "Additive Gaussian";
 k["Problem"]["Likelihood"]["Reference Data"] = getReferenceData();

 // Configuring the problem's variables and their prior distributions
 k["Variables"][0]["Name"] = "a";
 k["Variables"][0]["Type"] = "Computational";
 k["Variables"][0]["Prior Distribution"]["Type"] = "Uniform";
 k["Variables"][0]["Prior Distribution"]["Minimum"] = -5.0;
 k["Variables"][0]["Prior Distribution"]["Maximum"] = +5.0;
 k["Variables"][0]["Lower Bound"] = -5.0;
 k["Variables"][0]["Upper Bound"] = +5.0;

 k["Variables"][1]["Name"] = "b";
 k["Variables"][1]["Type"] = "Computational";
 k["Variables"][1]["Prior Distribution"]["Type"] = "Uniform";
 k["Variables"][1]["Prior Distribution"]["Minimum"] = -5.0;
 k["Variables"][1]["Prior Distribution"]["Maximum"] = +5.0;
 k["Variables"][1]["Lower Bound"] = -5.0;
 k["Variables"][1]["Upper Bound"] = +5.0;

 k["Variables"][2]["Name"] = "Sigma";
 k["Variables"][2]["Type"] = "Statistical";
 k["Variables"][2]["Prior Distribution"]["Type"] = "Uniform";
 k["Variables"][2]["Prior Distribution"]["Minimum"] = 0.0;
 k["Variables"][2]["Prior Distribution"]["Maximum"] = +5.0;
 k["Variables"][2]["Lower Bound"] = 0.0;
 k["Variables"][2]["Upper Bound"] = +5.0;

 // Configuring Solver
 k["Solver"]["Type"] = "CMAES";
 k["Solver"]["Sample Count"] = 12;

 // General Settings
 k["General"]["Max Generations"] = 1000;
 k["General"]["Results Output"]["Frequency"] = 5;
 k["General"]["Console Output"]["Frequency"] = 5;

 // Setting Model
 k.setModel(model);

 // Running Korali
 k.run();
}

File: tutorials/a3-bayesian-inference/run-cmaes.py
#!/usr/bin/env python3

# In this example, we demonstrate how Korali find the variable values
# that maximize the posterior in a bayesian problem where the likelihood
# is calculated by providing reference data points and their objective values.

# Importing the computational model
import sys
sys.path.append('./model')
from model import *
import korali

k = korali.initialize()

# Setting up the reference likelihood for the Bayesian Problem
k["Problem"]["Type"] = "Evaluation/Bayesian/Inference/Reference"
k["Problem"]["Likelihood Model"] = "Additive Normal"
k["Problem"]["Reference Data"] = getReferenceData()
k["Problem"]["Computational Model"] = lambda sampleData: model(sampleData, getReferencePoints())

# Configuring CMA-ES parameters
k["Solver"]["Type"] = "Optimizer/CMAES"
k["Solver"]["Population Size"] = 24 
k["Solver"]["Termination Criteria"]["Max Generations"] = 100

# Configuring the problem's random distributions
k["Distributions"][0]["Name"] = "Uniform 0"
k["Distributions"][0]["Type"] = "Univariate/Uniform"
k["Distributions"][0]["Minimum"] = -5.0
k["Distributions"][0]["Maximum"] = +5.0

k["Distributions"][1]["Name"] = "Uniform 1"
k["Distributions"][1]["Type"] = "Univariate/Uniform"
k["Distributions"][1]["Minimum"] = -5.0
k["Distributions"][1]["Maximum"] = +5.0

k["Distributions"][2]["Name"] = "Uniform 2"
k["Distributions"][2]["Type"] = "Univariate/Uniform"
k["Distributions"][2]["Minimum"] = 0.0
k["Distributions"][2]["Maximum"] = +5.0

# Configuring the problem's variables
k["Variables"][0]["Name"] = "a"
k["Variables"][0]["Bayesian Type"] = "Computational"
k["Variables"][0]["Prior Distribution"] = "Uniform 0"
k["Variables"][0]["Initial Mean"] = +0.0
k["Variables"][0]["Initial Standard Deviation"] = +1.0

k["Variables"][1]["Name"] = "b"
k["Variables"][1]["Bayesian Type"] = "Computational"
k["Variables"][1]["Prior Distribution"] = "Uniform 1"
k["Variables"][1]["Initial Mean"] = +0.0
k["Variables"][1]["Initial Standard Deviation"] = +1.0

k["Variables"][2]["Name"] = "Sigma"
k["Variables"][2]["Bayesian Type"] = "Statistical"
k["Variables"][2]["Prior Distribution"] = "Uniform 2"
k["Variables"][2]["Initial Mean"] = +2.5
k["Variables"][2]["Initial Standard Deviation"] = +0.5

# General Settings
k["Results Output"]["Frequency"] = 5
k["Console Output"]["Frequency"] = 5

# Running Korali
k.run()
